substrate/frame/revive/src/precompiles/builtin/storage.rs

```diff
@@ -0,0 +1,193 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::{
+	precompiles::{BuiltinAddressMatcher, BuiltinPrecompile, Error, Ext},
+	storage::WriteOutcome,
+	vm::RuntimeCosts,
+	Config, Key,
+};
+use alloc::vec::Vec;
+use alloy_core::sol_types::SolValue;
+use core::{marker::PhantomData, num::NonZero};
+use pallet_revive_uapi::{precompiles::storage::IStorage, StorageFlags};
+use sp_core::hexdisplay::AsBytesRef;
+
+pub struct Storage<T>(PhantomData<T>);
+
+impl<T: Config> BuiltinPrecompile for Storage<T> {
+	type T = T;
+	type Interface = IStorage::IStorageCalls;
+	const MATCHER: BuiltinAddressMatcher =
+		BuiltinAddressMatcher::Fixed(NonZero::new(0x901).unwrap());
+	const HAS_CONTRACT_INFO: bool = false;
+
+	fn call(
+		_address: &[u8; 20],
+		input: &Self::Interface,
+		env: &mut impl Ext<T = Self::T>,
+	) -> Result<Vec<u8>, Error> {
+		// Benchmarks call the pre-compile functions directly, without the delegate
+		// call overhead. The `delegate_call` overhead is benchmarked individually.
+		#[cfg(not(feature = "runtime-benchmarks"))]
+		assert!(env.is_delegate_call(), "Storage precompile can only be called via delegate call");
```

Comment:
We can't panic here. A contract can just do a normal call to trigger a runtime panic. Need to return an error.